lib/semantics/check-omp-structure.cc

```diff
@@ -14,9 +14,82 @@
 
 #include "check-omp-structure.h"
 #include "../parser/parse-tree.h"
+#include <variant>
 
 namespace Fortran::semantics {
 
+bool OmpStructureChecker::HasInvalidCloselyNestedRegion(
+    const parser::CharBlock &source, const OmpDirectiveEnumType &set) {
+  if (ompContext_.size() && set.test(GetOmpContext().currentDirectiveEnum)) {
+    context_.Say(source,
+        "A worksharing region may not be closely nested inside a "
+        "worksharing, explicit task, taskloop, critical, ordered, atomic, or "
+        "master region."_err_en_US);
+    return true;
+  }
+  return false;
+}
+
+void OmpStructureChecker::CheckAllowed(const OmpClauseEnum &type) {
+  if (ompContext_.empty()) {
+    return;
+  }
+  bool f{true};
+  if (!GetOmpContext().allowedClauses.test(type)) {
+    context_.Say(GetOmpContext().currentClauseSource,
+        "'%s' clause is not allowed on the %s directive"_err_en_US,
+        parser::ToUpperCaseLetters(EnumToString(type)),
+        parser::ToUpperCaseLetters(
+            GetOmpContext().currentDirectiveSource.ToString()));
+    f = false;
+  }
+  if (!GetOmpContext().allowedOnceClauses.empty() &&
+      GetOmpContext().allowedOnceClauses.test(type) &&
+      GetOmpContext().seenClauses.test(type)) {
+    context_.Say(GetOmpContext().currentClauseSource,
+        "At most one '%s' clause can appear on the %s directive"_err_en_US,
+        parser::ToUpperCaseLetters(EnumToString(type)),
+        parser::ToUpperCaseLetters(
+            GetOmpContext().currentDirectiveSource.ToString()));
+    f = false;
+  }
+  if (f) {
+    SetOmpContextSeen(type);
+  }
+}
+
+void OmpStructureChecker::Enter(const parser::OpenMPLoopConstruct &x) {
+  bool f{false};
+  const auto &dir{std::get<parser::OmpLoopDirective>(x.t)};
+  if (parser::ToUpperCaseLetters(dir.source.ToString()) == "DO") {
+    f = HasInvalidCloselyNestedRegion(dir.source,
+        {OmpDirectiveEnum::DO, OmpDirectiveEnum::SECTIONS,
+            OmpDirectiveEnum::SINGLE, OmpDirectiveEnum::WORKSHARE});
+  }
+  if (!f) {
+    OmpContext ct{dir.source};
+    ompContext_.push_back(ct);
```

Comment:
It might be better to push a context even in the error case. That would allow you to eliminate some of the checks for an empty stack, for example.

Reply:
Addressed in commit [cb89744](cb89744746186bcf120d8ca62a791e2778366b9b).